state: initial draft of a mock chain state

hackaugusto opened the pull request:
Opening the PR to expose the design and get feedback.

hackaugusto commented on objects/src/state.rs:
```diff
@@ -0,0 +1,84 @@
+        let block_num = self.blocks.len();
+        let chain_root = self.chain.mmr().accumulator();
+
+        let new = BlockHeader::new(
+            prev_hash,
+            block_num,
+            chain_root,
+            account_root: Digest,
+            nullifier_root: Digest,
+            note_root: Digest,
+            batch_root: Digest,
+            proof_hash: Digest,
+            version: Felt,
+            timestamp: Felt,
+        );
```
The question is:

- what is the preferred way of maintaining a account / nullifier / note trees to update the merkle tree and set these values?

bobbinth commented on objects/src/state.rs:
```diff
@@ -0,0 +1,84 @@
+use crate::{AccountId, BlockHeader, ChainMmr, Digest, Vec};
+use crypto::merkle::{DefaultMerkleStore, TieredSmt};
+
+/// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+#[derive(Default, Debug, Clone)]
+pub struct State {
+    /// An append-only structure used to represent the history of blocks produced for this chain.
+    chain: ChainMmr,
+
+    /// Storage for the historical state of the chain.
+    ///
+    /// This store contains historical data for accounts, nullifiers, notes, and batches.
+    data: DefaultMerkleStore,
+
+    /// History of produced blocks.
+    blocks: Vec<BlockHeader>,
+
+    /// Tree containing the latest [Nullifier] tree.
+    nullifier_db: TieredSmt,
+
+    /// The latest block being produced
+    pending: PendingBlock,
+}
```
A few comments on this:

1. We should probably have a vector of all current account states here - e.g., `Vec<Account>`. Eventually, we'll replace it with something that emulates account DB - but for now just a vector should do. For the user standpoint, it might be useful to be able to get reference to an account by either ID or maybe just index.
2. Similarly, we should probably have a vector of all notes in the chain - e.g., `Vec<Note>` and then allow the user to get a note either by hash or by index (or maybe something else).
3. Having `PendingBlock` be a part of this struct feels a bit weird to me. When this struct is instantiated from a file, I'm not sure we need to have a "pending block". So, I wonder if this part can be refactored so that we do block construction externally.
4. I think having `data` as generic storage is fine for now but I also think that it's very likely we'll have to refactor this in the future to make separation between blocks much more clear. One reason for this is that we might want to get a list of all notes created in a given block, or a list of all accounts updated in a given block etc. And I think currently, this might be non-trivial.

Lastly, a minor nit: I wonder if the name `State` is right for this struct. It feel a bit too generic. Maybe something like `Chain` or `Mock` or `MockChain` would be better?

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,247 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        #[cfg_attr(feature = "serde", serde(skip))]
+        objects: Objects<R>,
+
+        // Objects that have been created and are waiting for a block.
+        //
+        // These objects will become available once the block is sealed.
+        #[cfg_attr(feature = "serde", serde(skip))]
+        pending_objects: Objects<R>,
```
skipping these because of the comment  https://github.com/0xPolygonMiden/miden-base/pull/211#discussion_r1308445952

> Having PendingBlock be a part of this struct feels a bit weird to me. When this struct is instantiated from a file, I'm not sure we need to have a "pending block". So, I wonder if this part can be refactored so that we do block construction externally.

hackaugusto commented on objects/src/builder/mod.rs:
```diff
@@ -153,13 +159,6 @@ impl<T: Rng> NonFungibleAssetBuilder<T> {
         Ok(Self { details_builder })
     }
 
-    pub fn with_details(
-        &mut self,
-        details: &NonFungibleAssetDetails,
-    ) -> Result<NonFungibleAsset, AssetError> {
-        NonFungibleAsset::new(&details)
-    }
```
It doesn't make a lot of sense to have this constructor, if the user has that object they can call `NonFungibleAsset::new` directly. This constructor would also make the use of `rng` totally unecessary here.

hackaugusto commented on objects/src/assets/nonfungible.rs:
```diff
@@ -69,6 +72,11 @@ impl NonFungibleAsset {
     }
 
+    /// Return ID of the faucet which issued this asset.
+    pub fn faucet_id(&self) -> AccountId {
+        AccountId::new_unchecked(self.0[FAUCET_ID_POS])
+    }
```
Let me know if this should be changed to a checked version instead.

bobbinth replied:
I think this is fine as is.

hackaugusto commented on objects/src/builder/mod.rs:
```diff
@@ -230,11 +230,25 @@ impl<T: Rng> AccountBuilder<T> {
         self
     }
 
+    pub fn add_assets<I: IntoIterator<Item = Asset>>(mut self, assets: I) -> Self {
+        for asset in assets.into_iter() {
+            self.assets.push(asset);
+        }
+        self
+    }
+
     pub fn add_storage_item(mut self, item: StorageItem) -> Self {
         self.storage_builder.add_item(item);
         self
     }
 
+    pub fn add_storage_items<I: IntoIterator<Item = StorageItem>>(mut self, items: I) -> Self {
+        for item in items.into_iter() {
+            self.storage_builder.add_item(item);
+        }
+        self
+    }
```
Added these versions because this is a owning builder, doing this in the caller makes for ugly code.

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,247 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        }
+
+        /// Creates a [Account] and add to the list of pending objects.
+        pub fn new_account<C, S, A>(
+            &mut self,
+            code: C,
+            storage: S,
+            assets: A,
+            immutable: bool,
+            on_chain: bool,
+        ) -> Result<AccountId, AccountError>
```
It is a bit unfortunate that we have to pass all of these arguments, since the builder has internal state for it already. But if we want to configure the type of account created this seems the best option.

bobbinth replied:
Another option (or rather an addition to this method) is to have some number of more specialized account constructors - e.g., `new_basic_wallet`.

hackaugusto replied:
added

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,247 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+
+            self.blocks.push(new);
+            self.chain.mmr_mut().add(new.hash());
+            self.objects.extend(&mut self.pending_objects);
```
I haven't added any validation code because all objects are created by the builders, which should never generate conflicting IDs because of the RNG.

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,248 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            Ok(faucet_id)
+        }
+
+        /// Creates a [AccountId] with type [AccountType::NonFungibleFaucet] and add to the list of
+        /// pending objects.
+        pub fn new_nonfungible_account(
```
nit: should this be `new_nonfungible_faucet`?

hackaugusto replied:
changed

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,248 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        }
+
+        /// Creates [FungibleAsset] from the fungible faucet at position `asset_pos`.
+        pub fn new_fungible_asset(
+            &mut self,
+            asset_pos: usize,
+            amount: u64,
+        ) -> Result<FungibleAsset, AssetError> {
+            self.objects.fungible_faucets[asset_pos].1.amount(amount)?.build()
+        }
+
+        /// Creates [NonFungibleAsset] from the nonfungible faucet at position `asset_pos`.
+        pub fn new_nonfungible_asset(
+            &mut self,
+            asset_pos: usize,
+        ) -> Result<NonFungibleAsset, AssetError> {
+            self.objects.nonfungible_faucets[asset_pos].1.build()
+        }
```
nit: should `asset_pos` actually be `faucet_pos`?

Also, is there a reason not to return just `Asset` from both of this functions?

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,248 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifier_db: TieredSmt,
+
+        // RNG used to seed builders.
+        //
+        // This is used to seed the [AccountBuilder] and the [NonFungibleAssetBuilder].
+        rng: R,
```
question: how is `rng` going to be serialized? (mostly just curious)

hackaugusto replied:
The RNGs derive serde's traits:

https://github.com/rust-random/rand/blob/f3dd0b885c4597b9617ca79987a0dd899ab29fcb/src/rngs/xoshiro128plusplus.rs#L23-L26

bobbinth reviewed (commented):
Thank you! Looks good! I left a few minor comments inline.

hackaugusto commented on objects/src/builder/mod.rs:
```diff
@@ -59,8 +119,8 @@ impl<T: Rng> AccountIdBuilder<T> {
-    pub fn code_root(&mut self, code_root: Digest) -> &mut Self {
-        self.code_root = code_root;
+    pub fn code<C: AsRef<str>>(&mut self, code: C) -> &mut Self {
+        self.code = code.as_ref().to_string();
```
Changed it to receive the code instead of the digest. It should be easier to use this way, because the user doesn't have to instantiate the `assembler` themselves.

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,404 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        pub fn build_notes_tree(&self) -> Result<SimpleSmt, MerkleError> {
+            let mut entries = Vec::with_capacity(self.notes.len() * 2);
+
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2) as u64;
+                (tree_index, note.hash().into())
+            }));
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2 + 1) as u64;
+                (tree_index, note.metadata().into())
+            }));
```
Note: These indexes starts from `0` for each `BlockHeader`.

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            let block_num: u64 = self.blocks.len().try_into().expect("usize to u64 failed");
+            let block_num: Felt = block_num.into();
+
+            for account in self.pending_objects.accounts.iter() {
+                let id: Felt = account.id().into();
+                self.accounts.update_leaf(id.as_int(), account.hash().into())?;
+            }
```
Note: This is building on top of the tree with the account state from the previous blocks.

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            self.blocks.push(header);
+            self.chain.mmr_mut().add(header.hash());
+            self.objects.update_with(&mut self.pending_objects, header, &notes);
+            self.data.extend(self.accounts.inner_nodes());
```
It is a bit unfortunate that we have to iterate over the complete tree on every block. But this should be fine for the this usecase.

bobbinth replied:
Yes - I think this is fine for now. At some point later, we'll probably have a separate tree for the account DB (similar to how we handle nullifiers now).

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+    impl<R: Rng> Objects<R> {
+        pub fn new() -> Self {
+            Self {
+                accounts: vec![],
+                fungible_faucets: vec![],
+                nonfungible_faucets: vec![],
```
Is the idea to store all historical states here as well? If we don't need to store the historical states, then these could be `BTreeMap`'s.

hackaugusto replied:
I was adding the historical state to the merkle store, so this would have a single entry per item.

I didn't see an advantage to using a map, but we can change it if you want. What key do you want to use for it?

bobbinth replied:
I was thinking the key would be account ID and this way we naturally guarantee that we can't have more than one account with the same ID in the set.

hackaugusto replied:
how would someone use the account then? if the data is created in a binary, and we only have the JSON file to load the data from, the `AccountId` is long gone. Iterating over the keys would not give the same order as the one used in during creation, so people would need to copy/paste the id meaning the data creating a dependency among them.

hackaugusto replied:
> Wouldn't the full BTreeMap be serialized in the JSON file? And the order should be the same as the BTreeMap would be sorted by the numeric value of account ID.

The account id is effectively a random value, so that means there is a total order, but that it is a total order of random values. If the initial chain has an account created with initial assets, and the tests wants to use that, it would need to search for the account instead of using the creation order.

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        /// An append-only structure used to represent the history of blocks produced for this chain.
+        chain: ChainMmr,
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
```
I wonder whether we actually need to store historical data for accounts and nullifiers. I guess it doesn't hurt, but I'm also not sure what we'd use it for.

Regarding notes: since we attach inclusion paths to each note when we seal a block, it seems like this data may be redundant (i.e., if we need to get a note inclusion path, we'd go to the note directly).

bobbinth commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifiers: TieredSmt,
+
+        /// Tree containing the latest hash of each account.
+        accounts: SimpleSmt,
```
Maybe leave at `TODO` here to indicate that this should be updated in the future to a Tiered SMT with 64-bit keys.

hackaugusto replied:
added

bobbinth reviewed (commented):
Looks good! Thank you! I left a few small comments inline. A couple of other comments:

- A small nit: it would be good to group functions into sections add section delimiters etc. (similar to how we do in other files) to improve readability a bit.
- I didn't see a method for updating existing accounts. How are you thinking of doing this?

hackaugusto commented on objects/src/mock/chain.rs:
```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+
+        /// Creates a [SimpleSmt] tree from the `notes`.
+        ///
+        /// The root of this tree is a commitment to [Note]s included in the block. The commitment
+        /// is not for the fields in the [Note] structure, but for the [NoteMetadata] and for a
+        /// subset of the [Note]'s attributes (i.e. recipient + vault).
```
changed

hackaugusto commented:
> it would be good to group functions into sections add section delimiters etc. (similar to how we do in other files) to improve readability a bit.

added

> I didn't see a method for updating existing accounts. How are you thinking of doing this?

added a `account_mut` to get a reference and change the account state.

bobbinth reviewed (approved):
Looks good! Thank you!
